### ISSUE TITLE
App Terminate - Don't Exit?

### ISSUE THREAD

**Issue by joseph-montanez** [atk4/ui#506]
This is a hard rock to swallow but App->terminate using PHP's exit is too harsh. I am integrating with Laravel and exit is used in a few places. Laravel cannot properly unload for things like cookies. A workaround I have right now is throwing a custom `exception` and then catching it, bypassing the rest of ATK UI logic, like `exit`. I know for anyone depending on ATK UI to exit the entire script, this method will not work for them. As there may be additional logic outside of ATK they are expecting to **not** run. I think you can still keep the exit logic, but have a flag to bypass an `exit` command. 

Below is an example I've done to override some App methods to make this work, but it does not have have a flag to keep the old logic. I think App->redirect using exit is fine, because its optional, I can use the built-in framework's redirect, but App->terminate is not optional.

```php
<?php

namespace App\Ui;

class App extends \atk4\ui\App
{
    /**
     * Runs app and echo rendered template.
     */
    public function run()
    {
        try {
            $this->run_called = true;
            $this->hook('beforeRender');
            $this->is_rendering = true;

            // if no App layout set
            if (!isset($this->html)) {
                throw new Exception(['App layout should be set.']);
            }

            $this->html->template->set('title', $this->title);
            $this->html->renderAll();

            $this->html->template->appendHTML('HEAD', $this->html->getJS());
            $this->is_rendering = false;
            $this->hook('beforeOutput');

            if (isset($_GET['__atk_callback']) && $this->catch_runaway_callbacks) {
                $this->terminate('!! Callback requested, but never reached. You may be missing some arguments in '.$_SERVER['REQUEST_URI']);
            }

            return $this->html->template->render();
        } catch (TerminatedException $e) {
            return $e->output ? $e->output : '';
        }
    }

    /**
     * Constructor.
     *
     * @param array $defaults
     */
    public function __construct($defaults = [])
    {
        $this->app = $this;

        // Process defaults
        if (is_string($defaults)) {
            $defaults = ['title' => $defaults];
        }

        if (isset($defaults[0])) {
            $defaults['title'] = $defaults[0];
            unset($defaults[0]);
        }

        /*
        if (is_array($defaults)) {
            throw new Exception(['Constructor requires array argument', 'arg' => $defaults]);
        }*/
        $this->setDefaults($defaults);
        /*

        foreach ($defaults as $key => $val) {
            if (is_array($val)) {
                $this->$key = array_merge(isset($this->$key) && is_array($this->$key) ? $this->$key : [], $val);
            } elseif (!is_null($val)) {
                $this->$key = $val;
            }
        }
         */

        // Set up template folder
        $this->template_dir = __DIR__.'/../../vendor/atk4/ui/template/'.$this->skin;

        // Set our exception handler
        if ($this->catch_exceptions) {
            set_exception_handler(function ($exception) {
                return $this->caughtException($exception);
            });
        }

        if (!$this->_initialized) {
            //$this->init();
        }

        if ($this->fix_incompatible) {
            if (PHP_MAJOR_VERSION >= 7) {
                set_error_handler(function ($errno, $errstr) {
                    return strpos($errstr, 'Declaration of') === 0;
                }, E_WARNING);
            }
        }

        // Always run app on shutdown
        if ($this->always_run) {
            if ($this->_cwd_restore) {
                $this->_cwd_restore = getcwd();
            }

            register_shutdown_function(function () {
                if (is_string($this->_cwd_restore)) {
                    chdir($this->_cwd_restore);
                }

                if (!$this->run_called) {
                    try {
                        $this->run();
                    } catch (TerminatedException $e) {
                        //-- This is okay!
                    } catch (\Exception $e) {
                        $this->caughtException($e);
                    }
                }
                exit;
            });
        }

        // Set up UI persistence
        if (!isset($this->ui_persistence)) {
            $this->ui_persistence = new \atk4\ui\Persistence\UI();
        }
    }


    /**
     * Will perform a preemptive output and terminate. Do not use this
     * directly, instead call it form Callback, jsCallback or similar
     * other classes.
     *
     * @param string $output
     *
     * @throws \App\Ui\TerminatedException
     */
    public function terminate($output = null)
    {
        $this->run_called = true; // prevent shutdown function from triggering.

        throw new TerminatedException($output, 'Application terminated');
    }
}
```

**Comment by romaninsh** [comment]
Most integrations override terminate() method to properly shut down the framework. Can you try that?

**Comment by joseph-montanez** [comment]
@romaninsh Oh I did that, but I needed to override `App::__construct`, because I throw an exception instead of exit. `App::__construct` catches all exceptions and displays them on **XHR** requests. The __construct method has a lot of code that I don't feel I should be overriding, seems like any of that code can change between releases. I think throwing an exception is the only real solution, as its the only other *escape hatch* logic that works. This means other people going down this path will also be overriding the `App::__construct`.

> Integration with other Frameworks
> If you use Agile UI in conjunction with another framework, then you may be using a framework-specific App class, that implements tighter integration with the host application or full-stack framework.

So at this point I have overriden the following to work within Laravel:

**Fixing Exit and Stop Laravel from Throwing 500 errors**
- **App::terminate** - Throws custom exception with deferred output in exception
- **App::__construct** - Catch the custom exception on XHR requests
- **App::run** - Catch custom exception during render process, also return string instead of echo

**Fixing URL to work with Routers**
- **App::url** - Stop adding .php to every link

**Fixing POST requests**
- **App::initIncludes** - Include jQuery code for CSRF

There isn't any documentation on integration with frameworks, so all of this has been me stumbling into ATK and seeing what why its breaking within Laravel. I feel like exiting, directly echoing, and no global way to add to XHR requests have been fairly complicated to discover. 

> And frameworks like Yii, Laravel, Symfony can use Agile UI through a simple integration.

Anyways you can close this if you feel like nothing need to be done here and overriding `App` is the right answer. I am very excited to leverage ATK within Laravel, and I'll be releasing a service package like `atk4/laravel-ad` to help others interested.

**Comment by romaninsh** [comment]
construct will only react to uncaught exceptions. Everything else works as normal. Also you can set 

$catch_exceptions = false (https://github.com/atk4/ui/blob/develop/src/App.php#L46)

The package would be great, and the reason why there is no integration documentation is because I simply cannot anticipate all the requirements from the frameworks. We did major overhauls around AJAX urls when integrating wordpress and once you share your Laravel integration code we can fine-tune ATK side.

**Comment by joseph-montanez** [comment]
Cool beans. As for the `$catch_exceptions`, its not used here https://github.com/atk4/ui/blob/develop/src/App.php#L174 which is used to catch an exception in XHR requests and display them.

**Comment by romaninsh** [comment]
Ah yes you are correct. Stray exceptions are handled by "set_exception_handler" while execution of run() is fenced and invokes caughtException.

I don't mind modify logic and split it up evenly into multiple methods to help you with the overriding.

**Comment by romaninsh** [comment]
So far - really curious how your integration will turn out! Catch me in gitter.im if you want to do some code lifting.

**Comment by PhilippGrashoff** [comment]
should be fixed by #790

**Comment by abbadon1334** [comment]
probably solved with #790 
with flag `$app->call_exit = false` working like is described by @joseph-montanez first message.
